File: src/openfl/net/URLLoader.nim
import ./URLRequest
import ./URLLoaderDataFormat
import ./SharedObject
import ../events/EventDispatcher
import ../events/Event
import ../events/IOErrorEvent
import std/httpclient
import std/asyncdispatch

type
  URLLoader* = ref object of EventDispatcher
    bytesLoaded*: int
    bytesTotal*: int
    data*: RootRef
    dataFormat*: URLLoaderDataFormat

proc newURLLoader*(request: URLRequest = nil): URLLoader =
  let self = URLLoader(
    bytesLoaded: 0,
    bytesTotal: 0,
    dataFormat: URLLoaderDataFormat.TEXT
  )
  self.initEventDispatcher()
  # TODO: load if request is not nil
  return self

proc load*(self: URLLoader, request: URLRequest) =
  let client = newAsyncHttpClient()

  proc handleLoad() {.async.} =
    try:
      let resp = await client.get(request.url)
      let body = await resp.body

      # RootRef can only be cast from a ref type. string is a value type.
      # We need to wrap it in a ref object if we want to store it in RootRef.
      self.data = StringWrapper(val: body)
      self.bytesLoaded = body.len
      self.bytesTotal = body.len
      discard self.dispatchEvent(newEvent(COMPLETE))
    except:
      let errorEvent = newIOErrorEvent(IO_ERROR)
      errorEvent.text = getCurrentExceptionMsg()
      discard self.dispatchEvent(errorEvent)
    finally:
      client.close()

  asyncCheck handleLoad()

proc close*(self: URLLoader) =
  # TODO: implement
  discard

File: kfl/src/main/kotlin/openfl/net/URLLoader.kt
package openfl.net

import io.ktor.client.*
import io.ktor.client.engine.cio.*
import io.ktor.client.request.*
import io.ktor.client.statement.*
import io.ktor.http.*
import kotlinx.coroutines.*
import openfl.events.*
import openfl.utils.ByteArray
import openfl.utils.ObjectPool

enum class URLLoaderDataFormat {
    BINARY,
    TEXT,
    VARIABLES
}

class URLLoader(request: URLRequest? = null) : EventDispatcher() {

    var bytesLoaded: Int = 0
    var bytesTotal: Int = 0
    var data: Any? = null
    var dataFormat: URLLoaderDataFormat = URLLoaderDataFormat.TEXT

    private val client = HttpClient(CIO)
    private var job: Job? = null

    init {
        if (request != null) {
            load(request)
        }
    }

    fun close() {
        job?.cancel()
        job = null
    }

    fun load(request: URLRequest) {
        val url = request.url ?: return
        dispatchEvent(Event(Event.OPEN))

        job = CoroutineScope(Dispatchers.IO).launch {
            try {
                val response: HttpResponse = client.request(url) {
                    method = HttpMethod.parse(request.method)
                    // TODO: Add headers and data from request
                }

                val status = response.status.value
                dispatchEvent(HTTPStatusEvent(HTTPStatusEvent.HTTP_STATUS, false, false, status))

                if (response.status.isSuccess()) {
                    when (dataFormat) {
                        URLLoaderDataFormat.TEXT -> {
                            data = response.bodyAsText()
                        }
                        URLLoaderDataFormat.BINARY -> {
                            // data = ByteArray(response.readRawBytes())
                        }
                        else -> {}
                    }
                    dispatchEvent(Event(Event.COMPLETE))
                } else {
                    dispatchEvent(IOErrorEvent(IOErrorEvent.IO_ERROR, text = "HTTP Error $status"))
                }
            } catch (e: Exception) {
                dispatchEvent(IOErrorEvent(IOErrorEvent.IO_ERROR, text = e.message ?: "Unknown error"))
            }
        }
    }
}
